emark/models.py

```diff
@@ -0,0 +1,74 @@
+import uuid
+
+from django.conf import settings
+from django.db import models
+from django.urls import reverse
+
+
+class Send(models.Model):
+    """Frozen replica of a sent email message."""
+
+    uuid = models.UUIDField(
+        unique=True, default=uuid.uuid4, editable=False, primary_key=True
+    )
+
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        on_delete=models.SET_NULL,
+        related_name="emark_emails",
+        null=True,
+    )
+    from_address = models.EmailField()
+    to_address = models.EmailField()
+    subject = models.TextField(max_length=998)  # RFC 2822
+    body = models.TextField()
+    html = models.TextField()
+    utm = models.JSONField(default=dict)
+    created_at = models.DateTimeField(auto_now_add=True)
+
+    def get_absolute_url(self):
+        return reverse("emark:email-detail", kwargs={"pk": self.pk})
+
+
+class ClientTrackingQueryset(models.QuerySet):
+    def create_for_request(self, request, **kwargs):
+        """Create a tracking record for the given request."""
+        return self.create(
+            headers=dict(request.headers),
+            ip_address=request.META.get("REMOTE_ADDR"),
+            utm={
+                key: value
+                for key, value in request.GET.items()
+                if key.startswith("utm_")
+            },
+            **kwargs,
+        )
+
+
+class ClientTrackingModelMixin(models.Model):
+    uuid = models.UUIDField(
+        unique=True, default=uuid.uuid4, editable=False, primary_key=True
+    )
+    headers = models.JSONField(default=dict)
+    ip_address = models.GenericIPAddressField()
+    utm = models.JSONField(default=dict)
+
+    objects = ClientTrackingQueryset.as_manager()
+
+    class Meta:
+        abstract = True
+
+
+class Click(ClientTrackingModelMixin):
+    """Record of a click on a link in an email."""
+
+    email = models.ForeignKey(Send, on_delete=models.CASCADE)
+    redirect_url = models.URLField()
```

Comment:
I managed to break this.
`URLField()` by default has max_length=200 but the link I had in my email is much longer:

<details>

```
status=500 method=GET path="/emark/c6f4b2f2-62f9-4b59-9c29-3b5100d48c6d/click?url=https%3A%2F%2Fvoiio-platfo-emark-trac-munpyq.herokuapp.com%2Faccounts%2Flogin%2FINq%3A%3A1qMP4r%3AZfeHkYAT9UsgjZ_mLo0gxIhQaQy3UfrJR3cpMBEcVgA%3Futm_source%3Dplatform%26utm_medium%3Demail%26utm_campaign%3DLOGIN_EMAIL%26next%3D%252Fdiscover%252F"


django.db.utils.DataError: value too long for type character varying(200)
```

</details>